File: trustdevice/src/main/java/cn/tongdun/mobrisk/core/collectors/MagiskCollector.java
package cn.tongdun.mobrisk.core.collectors;

import java.io.File;

import cn.tongdun.mobrisk.core.utils.Constants;
import cn.tongdun.mobrisk.core.utils.EnvUtils;

/**
 * @author collam
 * @date 2023/3/14
 */
public class MagiskCollector {
    public boolean detectMagisk(){
        return detectMagiskByFile() || detectMagiskByLastModified();
    }

    private boolean detectMagiskByLastModified(){
        File mountsFile = new File(Constants.MAGISK_MOUNTS_PATH);
        File mountInfoFile = new File(Constants.MAGISK_MOUNT_INFO_PATH);
        File mountStatsFile = new File(Constants.MAGISK_MOUNT_STATS_PATH);

        if(mountsFile.lastModified() == mountInfoFile.lastModified()
                && mountInfoFile.lastModified() == mountStatsFile.lastModified()){
            return false;
        }
        return true;
    }

    private boolean detectMagiskByFile(){
        return EnvUtils.fileInEnv("magisk");
    }
}

File: trustdevice/src/main/java/cn/tongdun/mobrisk/core/collectors/MagiskCollector.kt
package cn.tongdun.mobrisk.core.collectors

import cn.tongdun.mobrisk.core.tools.EnvUtils
import cn.tongdun.mobrisk.core.tools.JNIHelper

/**
 * @description:
 * @author: wuzuchang
 * @date: 2023/5/15
 */
interface MagiskInterface {
    fun detectMagisk(): Boolean
}

class MagiskCollector : MagiskInterface {

    override fun detectMagisk(): Boolean {
        return detectMagiskByFile() || detectMagiskByLastModified()
    }

    private fun detectMagiskByLastModified(): Boolean {
        return JNIHelper.call3()
    }

    private fun detectMagiskByFile(): Boolean {
        return EnvUtils.fileInEnv("magisk")
    }
}